crates/edr_provider/src/logger.rs

```diff
@@ -1,143 +1,177 @@
 // TODO: Remove once stubs have been implemented.
 #![allow(dead_code, clippy::unused_self)]
 
-use edr_eth::{
-    transaction::{SignedTransaction, TransactionRequest},
-    B256, U256,
-};
-use edr_evm::{trace::Trace, Bytecode, MineBlockResult, SyncBlock};
+use dyn_clone::DynClone;
+use edr_evm::{trace::Trace, ExecutableTransaction};
 
-use crate::ProviderError;
+use crate::{
+    data::{CallResult, EstimateGasFailure},
+    debug_mine::DebugMineBlockResult,
+    ProviderError,
+};
 
-/// A logger that handles all the logging of the RPC server.
-///
-/// The API follows this convention:
-/// - `log_*` methods accumulate logs.
-/// - `print_*` methods immediately print to stdout.
-pub struct Logger {
-    _is_enabled: bool,
-}
+pub trait Logger {
+    type BlockchainError;
 
-impl Logger {
-    /// Constructs a new instance.
-    pub fn new(is_enabled: bool) -> Self {
-        Self {
-            _is_enabled: is_enabled,
-        }
-    }
+    /// Whether the logger is enabled.
+    fn is_enabled(&self) -> bool;
 
-    /// Logs the result of auto-mining a block.
-    pub fn log_block_from_automine<BlockchainErrorT>(
-        &self,
-        _result: MineBlockResult<BlockchainErrorT>,
-        _contracts: Vec<Bytecode>,
-        _transaction_hash_to_highlight: &B256,
-    ) {
-    }
+    /// Sets whether the logger is enabled.
+    fn set_is_enabled(&mut self, is_enabled: bool);
 
-    /// Logs the result of mining a block.
-    pub fn log_mined_block<BlockchainErrorT>(
-        &self,
-        _result: MineBlockResult<BlockchainErrorT>,
-        _contracts: Vec<Bytecode>,
+    fn log_call(
+        &mut self,
+        spec_id: edr_eth::SpecId,
+        transaction: &ExecutableTransaction,
+        result: &CallResult,
```

Comment:
Nit: the arguments could be prefixed with `_` instead of doing it in the body

Reply:
The reason I did it in the body of default trait functions is because naming variables in a trait function with `_` prefix, results in the IDE copying those names when you auto-fill the function:

```rust
    fn log_estimate_gas_failure(
        &mut self,
        _spec_id: edr_eth::SpecId,
        _transaction: &ExecutableTransaction,
        _result: &EstimateGasFailure,
    ) {
    }
```

Doing it my way ensures that those parameters don't have the `_` prefix